# epigenomics

Especially targeted to functions related to the analysis of epigenomics data. It has functions to read, merge, denoise, ChIP seq data. It contains a lot of functions required for the [AML paper](github.com/JKBio/AMLproject).

## Available functions:

_in ./chipseq.py_

- dropWeirdChromosomes: 
- extractPairedSingleEndFrom: 
- findReplicates: 
- computeSingleEnd: 
- computePairedEnd: 
- bigWigFrom: 
- mergeBams: 
- loadPeaks: 
- pysam_computePeaksAt: 
- bedtools_computePeaksAt: 
- computePeaksAt: 
- simpleMergedPeaks: 
- findpeakpath: 
- mergeReplicatePeaks: 
- findAdditionalPeaks: 
- putInConscensus: 
- findAdditionalCobindingSignal: 
- fullDiffPeak: 
- diffPeak: 

## highly recommended packages

- Bedtools
- deepTools
- MACS2
- ROSE
- MEME
- ChromHMM